text/0000-stable-simd.md

```diff
@@ -0,0 +1,718 @@
+- Feature Name: `vendor_intrinsics`
+- Start Date: 2018-02-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The purpose of this RFC is to provide a framework for SIMD to be used on stable
+Rust. It proposes stabilizing x86-specific vendor intrinsics, but includes the
+scaffolding for other platforms as well as a future portable SIMD design (to be
+fleshed out in another RFC).
+
+# Motivation
+[motivation]: #motivation
+
+Stable Rust today does not typically expose all of the capabilities of the
+platform that you're running on. A notable gap in Rust's support includes SIMD
+(single instruction multiple data) support. For example on x86 you don't
+currently have explicit access to the 128, 256, and 512 bit registers on the
+CPU. LLVM is in general an excellent optimizing compiler and often attempts to
+make use of these registers (auto vectorizing code), but it unfortunately is
+still somewhat limited and doesn't express the full power of the various SIMD
+intrinsics.
+
+The goal of this RFC is to enable using SIMD intrinsics on stable Rust, and in
+general provide a means to access the architecture-specific functionality of
+each vendor. For example the AES intrinsics on x86 would also be made available
+through this RFC, not only the SIMD-related AVX intrinsics.
+
+Note that this is certainly not the first discussion to broach the topic of SIMD
+in Rust, but rather this has been an ongoing discussion for quite some time now!
+For example the [simd crate][simd-crate] started [long ago][simd-start], we've
+had [rfcs][simd-rfc], we've had a [lot][i1] of [discussions][i2] on internals,
+and the [stdsimd] crate has been implemented.
+
+This RFC draws from much of the historical feedback and design that we've done
+around SIMD in Rust and is targeted at providing path forward for using SIMD on
+stable Rust while allowing the compiler to change in the future and retain a
+stable interface.
+
+[simd-rfc]: https://github.com/rust-lang/rfcs/pull/1199
+[simd-crate]: https://github.com/rust-lang-nursery/simd
+[simd-start]: http://huonw.github.io/blog/2015/08/simd-in-rust/
+[stdsimd]: https://github.com/rust-lang-nursery/stdsimd
+[i1]: https://internals.rust-lang.org/t/getting-explicit-simd-on-stable-rust/4380
+[i2]: https://internals.rust-lang.org/t/whats-the-next-step-towards-the-stabilization-of-simd/5867
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Let's say you've just heard about this fancy feature called "auto vectorization"
+in LLVM and you want to take advantage of it. For example you've got a function
+like this you'd like to make faster:
+
+```rust
+pub fn foo(a: &[u8], b: &[u8], c: &mut [u8]) {
+    for ((a, b), c) in a.iter().zip(b).zip(c) {
+        *c = *a + *b;
+    }
+}
+```
+
+When [inspecting the assembly][asm1] you notice that rustc is making use of the
+`%xmmN` registers which you've read is related to SSE on your CPU. You know,
```

Comment:
Nit: xmm or ymm?

Reply:
xmm. ymm are 256 bit registers, and in this particular example, there are no target features enabled, so on x86_64 it will be limited to SSE2 instructions, which use 128 bit registers (xmm).

The example below enables the `avx2` feature, which permits AVX2 instructions, which use 256 bit registers (ymm).